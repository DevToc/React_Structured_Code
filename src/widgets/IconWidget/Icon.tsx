import SVG from 'react-inlinesvg';
import { Box } from '@chakra-ui/react';
import { IconApiResponse } from 'hooks/useIcon';
import { preProcessSvg, generateClipPath, getIconStyle } from './IconWidget.helpers';
import { FillDirection } from './IconWidget.types';

interface IconProps {
  iconConfig: {
    shapeColorOne: string;
    shapeColorTwo?: string;
    shapeFill: number;
    fillDirection: FillDirection;
    isMirrored?: boolean;
  };
  iconData: IconApiResponse;
}

export const Icon = ({ iconConfig, iconData }: IconProps) => {
  const { shapeColorOne, isMirrored = false, shapeColorTwo, shapeFill, fillDirection } = iconConfig;
  const { svg, viewBox, description, color } = iconData;

  const iconStyle = getIconStyle(isMirrored);
  const hasSecondaryColorFill = !!(typeof shapeFill === 'number' && shapeFill <= 100 && shapeFill > 0 && shapeColorTwo);
  const formattedSvg: string = preProcessSvg(svg, color);

  // could be added to iconData response?
  const title = 'icon';
  // seems to be "missing" for most icons
  const iconDescription = description || 'icon description';

  return (
    <Box w='100%' h='100%' position='absolute'>
      <SVG
        title={title}
        description={iconDescription}
        role='img'
        src={formattedSvg}
        fill={color ? '' : shapeColorOne}
        style={{ ...iconStyle, position: 'absolute' }}
        uniquifyIDs={true}
        {...(viewBox && { viewBox })}
      />
      {hasSecondaryColorFill && (
        <Box
          position='absolute'
          zIndex={1}
          w='100%'
          h='100%'
          clipPath={generateClipPath(shapeFill, fillDirection)}
          bottom='0'
        >
          <SVG
            title={title}
            description={iconDescription}
            role='img'
            src={formattedSvg}
            fill={shapeColorTwo}
            style={iconStyle}
            uniquifyIDs={true}
            {...(viewBox && { viewBox })}
          />
        </Box>
      )}
    </Box>
  );
};
