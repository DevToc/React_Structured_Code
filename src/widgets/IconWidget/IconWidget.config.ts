import { HANDLE } from 'constants/bounding-box';

const BOUNDING_BOX_CONFIG = {
  customHandle: HANDLE.CORNERS,
};
const MIN_NR_ICONS = 1;
const MAX_NR_ICONS = 21;
const NR_OF_ICONS = [...Array(MAX_NR_ICONS - MIN_NR_ICONS)].map((_, i) => i + MIN_NR_ICONS);

const DEFAULT_SHAPE_COLOR_ONE = '#0082FF';
const DEFAULT_SHAPE_COLOR_TWO = '#f64d2a';
const DEFAULT_NR_ICONS = 10;
const DEFAULT_GRID_ITEM_SIZE_PX = 42;
const DEFAULT_GRID_GAP_PX = 10;
const MAX_GAP_PX = 30;

export {
  BOUNDING_BOX_CONFIG,
  NR_OF_ICONS,
  MIN_NR_ICONS,
  MAX_NR_ICONS,
  DEFAULT_NR_ICONS,
  DEFAULT_GRID_ITEM_SIZE_PX,
  DEFAULT_GRID_GAP_PX,
  MAX_GAP_PX,
  DEFAULT_SHAPE_COLOR_ONE,
  DEFAULT_SHAPE_COLOR_TWO,
};
