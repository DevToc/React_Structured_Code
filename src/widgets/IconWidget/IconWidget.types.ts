import { Widget, AccessibleElement } from 'types/widget.types';

enum IconWidgetType {
  Grid = 'Grid',
  Single = 'Single',
}

enum FillDirection {
  TopDown = 'TopDown',
  LeftRight = 'LeftRight',
}

interface IconWidgetData extends Widget, AccessibleElement {
  iconId: string;
  shapeColorOne: string;
  isMirrored: boolean;
  shapeColorTwo: string;
  shapeFill: number;
  fillDirection?: FillDirection;

  // The icon widget can be a grid (same icon repeated) or a single icon
  type?: IconWidgetType;

  // grid properties
  numberOfIcons?: number;
  gridGapPx?: number;
  gridItemWidthPx?: number;
  gridItemHeightPx?: number;
}

interface IconStyle {
  height: string;
  width: string;
  transform?: string;
  transformOrigin?: string;
}

interface AdjustSvgViewBox {
  viewBox?: string;
  maxSize: number;
}

interface Dimensions {
  width: number;
  height: number;
}

export type { IconWidgetData, IconStyle, AdjustSvgViewBox, Dimensions };
export { IconWidgetType, FillDirection };
