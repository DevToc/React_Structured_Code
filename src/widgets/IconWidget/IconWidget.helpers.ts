import { WidgetType } from 'types/widget.types';
import { WidgetId } from 'types/idTypes';
import {
  IconStyle,
  AdjustSvgViewBox,
  Dimensions,
  IconWidgetData,
  IconWidgetType,
  FillDirection,
} from './IconWidget.types';
import { VERSION } from './IconWidget.upgrade';
import {
  DEFAULT_NR_ICONS,
  DEFAULT_GRID_ITEM_SIZE_PX,
  DEFAULT_GRID_GAP_PX,
  DEFAULT_SHAPE_COLOR_ONE,
  DEFAULT_SHAPE_COLOR_TWO,
} from './IconWidget.config';

const FILL_SVG_REGEX = /fill=".*?"/g;
const removeAllFill = (svgHtml: string): string => svgHtml?.replace(FILL_SVG_REGEX, '');

const preProcessSvg = (svgHtml: string, color: number): string => (color ? svgHtml : removeAllFill(svgHtml));

const generateClipPath = (fill: number, fillDirection: FillDirection): string => {
  const clipFill = 100 - fill;

  if (fillDirection === FillDirection.TopDown) return `inset(0% 0% ${clipFill}% 0%)`;
  if (fillDirection === FillDirection.LeftRight) return `inset(0% ${clipFill}% 0% 0%)`;

  return '';
};

const getIconStyle = (isMirrored: boolean): IconStyle => {
  const iconStyle: IconStyle = { width: '100%', height: '100%' };

  if (isMirrored) {
    iconStyle.transform = 'scale(-1, 1)';
    iconStyle.transformOrigin = 'center';
  }

  return iconStyle;
};

const getIconGridId = (widgetId: WidgetId): string => `iconGrid${widgetId}`;
const getIconGridEl = (widgetId: WidgetId): HTMLElement => document.getElementById(getIconGridId(widgetId))!;

/**
 * Get the height of the grid based on the number of icons.
 * This is done by cloning the widget and adding the new number of icons.
 * This is then measured and the clone is removed.
 * This is done because the grid height is not known until the icons are rendered.
 *
 * @param widgetId
 * @param numberOfIcons
 * @return {number} - new height of the grid
 */
const getHeightOfGrid = (widgetId: WidgetId, numberOfIcons: number) => {
  const clonedWidget = document.getElementById(widgetId)!.cloneNode(true) as HTMLElement;
  const iconGridEl = clonedWidget.querySelector('div')! as HTMLElement;
  const firstChildClone = iconGridEl.firstChild!.cloneNode(true);

  // remove all children
  while (iconGridEl.firstChild) {
    iconGridEl.removeChild(iconGridEl.firstChild);
  }

  // add the new number of icons
  for (let i = 0; i < numberOfIcons; i++) {
    iconGridEl.appendChild(firstChildClone.cloneNode(true));
  }

  // add the cloned widget to the DOM
  clonedWidget.style.visibility = 'hidden';
  document.body.appendChild(clonedWidget);

  // measure the height
  const newHeight = iconGridEl.clientHeight;

  // remove the clone
  document.body.removeChild(clonedWidget);

  return newHeight;
};

/**
 * Get the height of the grid based on the grid gap.
 * This is done by cloning the widget and setting the new grid gap.
 * This is then measured and the clone is removed.
 * This is done because the grid height is not known until the icons with gap are rendered.
 *
 * @param widgetId
 * @param gapPx
 * @return {number} - new height of the grid
 */
const getHeightOfGap = (widgetId: WidgetId, gapPx: number) => {
  const clonedWidget = document.getElementById(widgetId)!.cloneNode(true) as HTMLElement;
  const iconGridEl = clonedWidget.querySelector('div')! as HTMLElement;

  // set new grid gap
  iconGridEl.style.gap = `${gapPx}px`;

  // add the cloned widget to the DOM
  clonedWidget.style.visibility = 'hidden';
  document.body.appendChild(clonedWidget);

  // measure the height
  const newHeightPx = iconGridEl.clientHeight;

  // remove the clone
  document.body.removeChild(clonedWidget);

  return newHeightPx;
};

/**
 * Get the height of the grid widget for a new icon size.
 * This is done by cloning the widget and setting the new icon size for all icons.
 * This is then measured and the clone is removed.
 * This is done because the height is not known until the icons with the new dimensions are rendered.
 *
 * @param widgetId
 * @param gridIconWidth - new width of the grid icon
 * @param gridIconHeight - new height of the grid icon
 * @return {{ width: number, height: number}} - new height of the grid
 */
const getGridIconSize = (widgetId: WidgetId, gridIconWidth: number, gridIconHeight: number) => {
  const clonedWidget = document.getElementById(widgetId)!.cloneNode(true) as HTMLElement;
  const iconGridEl = clonedWidget.querySelector('div')! as HTMLElement;

  // update size of all children
  const children = iconGridEl.children as HTMLCollectionOf<HTMLElement>;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    child.style.width = `${gridIconWidth}px`;
    child.style.height = `${gridIconHeight}px`;
  }

  // add the cloned widget to the DOM
  clonedWidget.style.visibility = 'hidden';
  document.body.appendChild(clonedWidget);

  // measure the height
  const newHeightPx = iconGridEl.clientHeight;
  const newWidthPx = iconGridEl.clientWidth;

  // remove the clone
  document.body.removeChild(clonedWidget);

  return { width: newWidthPx, height: newHeightPx };
};

/**
 * Convert svg and calculate the aspect ratio fit for the provided max size.
 *
 * @param viewBox - The icon meta data object
 * @param maxSize - max size for height / width
 * @return {Dimensions}
 */
const adjustSVGViewBox = ({ viewBox, maxSize }: AdjustSvgViewBox): Dimensions => {
  if (!viewBox) return { width: maxSize, height: maxSize };

  const viewBoxArr = viewBox.split(' ').map((view) => +view);
  const adjustRatio = maxSize / Math.max(viewBoxArr[2], viewBoxArr[3]);

  const width = adjustRatio * viewBoxArr[2];
  const height = adjustRatio * viewBoxArr[3];

  return { width, height };
};

/**
 * Get the icon grid array.
 * This is used to determine the fill color of each icon in the grid
 * The fill color is determined by the fill percentage and the number of icons.
 *
 * // 0 = 100% fill color one , 1 = 100% fill color two
 * E.g 55% fill w/ 10 icons -> [0, 0, 0, 0, 0.5, 1, 1, 1, 1, 1]
 * E.g 30% fill w/ 10 icons -> [0, 0, 0, 1, 1, 1, 1, 1, 1, 1]
 *
 * @param viewBox - The icon meta data object
 * @param maxSize - max size for height / width
 * @return {{ shapeFill: number; shapeColorOne: string; shapeColorTwo: string | undefined }[]}
 */
const getIconGridArr = (
  numberOfIcons: number,
  shapeFill: number,
  shapeColorOne: string,
  shapeColorTwo: string,
): { shapeFill: number; shapeColorOne: string; shapeColorTwo: string | undefined }[] => {
  const fillPercent = shapeFill / 100;
  const startFillIndex = fillPercent * numberOfIcons;

  // split the number into whole integer and float part
  // e.g. 5.5 -> [5, 5], 33.33 -> [33, 33],
  const [num, float] = startFillIndex
    .toString()
    .split('.')
    .map((n) => {
      if (n.length > 1 && n[0] === '0') return parseInt(n[1]) / 10;
      return parseInt(n);
    });

  const iconArr = [];
  for (let i = 0; i < numberOfIcons; i++) {
    let colorOne = i < startFillIndex ? shapeColorOne : shapeColorTwo;

    // default fill is 100% for all icons in the grid
    let fill = 100;
    // only partially filled icons need the secondary color
    let colorTwo = undefined;

    const hasPartialFill = i === num && float > 0;
    if (hasPartialFill) {
      // shape colors are reversed for partial filled icons
      colorTwo = shapeColorOne;
      colorOne = shapeColorTwo;

      // if the float is less than 1 (e.g. 0.1), then the fill is the float * 10
      if (!Number.isInteger(float) && float < 1) {
        fill = float * 10;
      } else {
        // get the first two numbers of the float part
        const fillPrecision = 2;
        const formatFloatString = float < 10 ? String(float).padEnd(fillPrecision, '0') : String(float);
        const partialFill = Number(formatFloatString.slice(0, fillPrecision));

        fill = partialFill;
      }
    }

    iconArr.push({ shapeColorOne: colorOne, shapeColorTwo: colorTwo, shapeFill: fill });
  }

  return iconArr;
};

/**
 * Generate default icon widget data
 *
 * @param iconId - icon id from the icon API
 * @param viewBox - icon meta data viewbox string (for the icons aspect ratio)
 * @param topPxOverride - top position override
 * @param leftPxOverride - left position override
 * @param iconConfig - config for overriding default icon widget data
 * @returns default icon widget data
 */
const generateDefaultData = (
  iconId: string,
  viewBox?: string,
  topPxOverride?: number,
  leftPxOverride?: number,
  iconConfig?: Partial<IconWidgetData>,
) => {
  const { width: widthPx, height: heightPx } = adjustSVGViewBox({ viewBox, maxSize: 300 });

  const widget = {
    widgetType: WidgetType.Icon,
    widgetData: {
      version: VERSION,
      topPx: topPxOverride || 0,
      leftPx: leftPxOverride || 0,
      rotateDeg: 0,
      iconId,
      widthPx: iconConfig?.widthPx || widthPx,
      heightPx: iconConfig?.heightPx || heightPx,
      shapeColorOne: iconConfig?.shapeColorOne || DEFAULT_SHAPE_COLOR_ONE,
      shapeColorTwo: iconConfig?.shapeColorTwo || DEFAULT_SHAPE_COLOR_TWO,
      shapeFill: iconConfig?.shapeFill || 0,
      isMirrored: false,
      altText: '',
      isDecorative: true,
      numberOfIcons: DEFAULT_NR_ICONS,
      gridGapPx: DEFAULT_GRID_GAP_PX,
      gridItemWidthPx: DEFAULT_GRID_ITEM_SIZE_PX,
      gridItemHeightPx: DEFAULT_GRID_ITEM_SIZE_PX,
      type: iconConfig?.type || IconWidgetType.Single,
      isHidden: iconConfig?.isHidden || false,
    },
  };

  return widget;
};

export {
  adjustSVGViewBox,
  getIconGridArr,
  generateDefaultData,
  getHeightOfGap,
  generateClipPath,
  getIconStyle,
  getHeightOfGrid,
  getIconGridId,
  getIconGridEl,
  preProcessSvg,
  getGridIconSize,
};
