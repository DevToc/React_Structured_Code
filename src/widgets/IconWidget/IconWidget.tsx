import { ReactElement, Suspense, useMemo } from 'react';
import { Box, Flex, Spinner } from '@chakra-ui/react';
import { useIcon, IconApiResponse } from 'hooks/useIcon';
import {
  useWidget,
  WidgetToolbar,
  WidgetBase,
  ReadOnlyWidgetBase,
  WidgetBaseProp,
  TransparentImg,
  useEditor,
} from 'widgets/sdk';
import { IconWidgetData, IconWidgetType, FillDirection } from './IconWidget.types';
import { DEFAULT_NR_ICONS, DEFAULT_GRID_ITEM_SIZE_PX, DEFAULT_GRID_GAP_PX } from './IconWidget.config';
import { getIconGridArr, getIconGridId } from './IconWidget.helpers';
import { Icon } from './Icon';
import { useIconResize } from './useIconResize';
import { useDynamicImport } from 'hooks/useDynamicImport';
import { WidgetComponentsMap } from 'widgets/components.lazy';

export const IconWidget = (): ReactElement => {
  const { customOnResize, customOnResizeStart, customOnResizeEnd } = useIconResize();
  const { openReplaceWidgetMenu } = useEditor();

  const onDoubleClick = () => openReplaceWidgetMenu();

  const ResponsiveComponent = (name: string): ReactElement => {
    const Component = useDynamicImport(name, WidgetComponentsMap);
    return (
      <Suspense fallback={<Spinner />}>
        <Component />
      </Suspense>
    );
  };

  const IconWidgetToolbarMenu = (): ReactElement => ResponsiveComponent('iconWidgetToolbar');

  return (
    <WidgetBase
      onDoubleClick={onDoubleClick}
      onResize={customOnResize}
      onResizeStart={customOnResizeStart}
      onResizeEnd={customOnResizeEnd}
    >
      <WidgetToolbar>
        <IconWidgetToolbarMenu />
      </WidgetToolbar>
      <IconWidgetBase />
    </WidgetBase>
  );
};

export const ReadOnlyIconWidget = ({ includeAltTextImg }: WidgetBaseProp): ReactElement => {
  return (
    <ReadOnlyWidgetBase>
      <IconWidgetBase isReadOnly />
      {includeAltTextImg && <TransparentImg />}
    </ReadOnlyWidgetBase>
  );
};

const IconWidgetBase = ({ isReadOnly = false }: { isReadOnly?: boolean }) => {
  const { iconId, type = IconWidgetType.Single } = useWidget<IconWidgetData>();
  const { data: iconData, error } = useIcon(iconId);

  return (
    <>
      {iconData && type === IconWidgetType.Grid && <IconGrid isReadOnly={isReadOnly} iconData={iconData} />}
      {iconData && type === IconWidgetType.Single && <SingleIcon iconData={iconData} />}
      {error && <>{error.message}</>}
    </>
  );
};

const IconGrid = ({ iconData, isReadOnly }: { iconData: IconApiResponse; isReadOnly: boolean }) => {
  const {
    shapeColorOne,
    shapeColorTwo,
    shapeFill,
    gridGapPx = DEFAULT_GRID_GAP_PX,
    fillDirection = FillDirection.LeftRight,
    gridItemWidthPx = DEFAULT_GRID_ITEM_SIZE_PX,
    gridItemHeightPx = DEFAULT_GRID_ITEM_SIZE_PX,
    numberOfIcons = DEFAULT_NR_ICONS,
    widgetId,
  } = useWidget<IconWidgetData>();

  const iconGridArr = useMemo(
    () => getIconGridArr(numberOfIcons, shapeFill, shapeColorOne, shapeColorTwo),
    [numberOfIcons, shapeFill, shapeColorOne, shapeColorTwo],
  );

  const iconStyle = { width: gridItemWidthPx + 'px', height: gridItemHeightPx + 'px' };
  const flexStyle = { gap: gridGapPx + 'px' };
  const iconGridId = isReadOnly ? '' : getIconGridId(widgetId);

  return (
    <Flex id={iconGridId} wrap='wrap' style={flexStyle}>
      {iconGridArr.map((gridIcon, i) => {
        const key = `${widgetId}-grid-${i}`;
        const { shapeColorOne, shapeColorTwo, shapeFill } = gridIcon;
        const iconConfig = { shapeColorOne, shapeColorTwo, shapeFill, fillDirection };

        return (
          <Box key={key} style={iconStyle} position='relative'>
            <Icon iconData={iconData} iconConfig={iconConfig} />
          </Box>
        );
      })}
    </Flex>
  );
};

const SingleIcon = ({ iconData }: { iconData: IconApiResponse }): ReactElement => {
  const {
    shapeColorOne,
    shapeColorTwo,
    shapeFill,
    isMirrored,
    fillDirection = FillDirection.TopDown,
  } = useWidget<IconWidgetData>();

  return (
    <Icon iconConfig={{ shapeColorOne, shapeColorTwo, shapeFill, isMirrored, fillDirection }} iconData={iconData} />
  );
};
