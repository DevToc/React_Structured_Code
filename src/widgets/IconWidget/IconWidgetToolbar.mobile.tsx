import { ReactElement, Suspense, useMemo } from 'react';
import {
  Box,
  Drawer,
  DrawerOverlay,
  DrawerContent,
  DrawerHeader,
  DrawerBody,
  Flex,
  IconButton,
  Text,
  useDisclosure,
} from '@chakra-ui/react';

import { useWindowSize } from 'hooks/useWindowSize';

import { useWidget } from 'widgets/sdk';
import { useIcon } from 'hooks/useIcon';
import { ColorPickerInContext } from 'modules/Editor/components/ColorPicker';
import { AltTextMenu } from 'modules/Editor/components/Toolbar/AltTextMenu/AltTextMenu';
import { SliderNumberInput } from 'modules/common/components/Input/SliderNumberInput';
import { IconWidgetData } from './IconWidget.types';

import { ReactComponent as MirrorIcon } from 'assets/icons/switch.svg';
import { ReactComponent as FillIcon } from 'assets/icons/drop.svg';
import { ReactComponent as CloseIcon } from 'assets/icons/close_circle.svg';

const COLOR_ICON_SIZE = 5;
const TRIGGER_BUTTON_SIZE = 8;

interface DefaultTriggerProps {
  onClick: () => void;
}
type IconStyle = 'fill' | 'border' | 'text';

interface Props extends DefaultTriggerProps {
  iconStyle: IconStyle;
  color: string;
}

export const IconWidgetToolbarMenuMobile = (): ReactElement => {
  const { iconId } = useWidget<IconWidgetData>();
  const { data: iconData } = useIcon(iconId);
  const { deviceWidth } = useWindowSize();

  const iconHasColorMenu = iconData && typeof iconData.color === 'number' && !iconData.color;

  return (
    <Flex w={deviceWidth - 15} p='10px' gap='8px' align='center' data-testid='basic-shape-widget-toolbar'>
      <IconMirrorMenu />
      {iconHasColorMenu && <IconColorMenu />}
      <AltTextMenuWrapper />
    </Flex>
  );
};

const AltTextMenuWrapper = () => {
  const { widgetId } = useWidget<IconWidgetData>();
  return (
    <Box //
      mr='5px'
      p='10px 10px'
      as='div'
      style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
    >
      <Box w='75px' as='div'>
        <AltTextMenu widgetId={widgetId} />
      </Box>
    </Box>
  );
};

const IconMirrorMenu = (): ReactElement | null => {
  const { isMirrored, updateWidget } = useWidget<IconWidgetData>();
  const onUpdateIconMirror = () => updateWidget({ isMirrored: !isMirrored });

  if (typeof isMirrored !== 'boolean') return null;

  return (
    <Box mr='5px' p='10px 10px' as='div' style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
      <IconButton
        mb='10px'
        onClick={onUpdateIconMirror}
        variant={'ghost'}
        isActive={isMirrored}
        size='sm'
        aria-label={'Mirror Icon'}
        icon={<MirrorIcon />}
      />
      <Box w='75px' as='div'>
        <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
          Mirror
        </Text>
      </Box>
    </Box>
  );
};

const IconColorMenu = (): ReactElement => {
  const { updateWidget, shapeFill, shapeColorOne, shapeColorTwo } = useWidget<IconWidgetData>();
  const { isOpen: isColorFillOpen, onToggle: toggleColorFill } = useDisclosure();
  const { isOpen: isFillColorOpen, onToggle: toggleFillColor } = useDisclosure();
  const { isOpen: isFillColorTwoOpen, onToggle: toggleFillColorTwo } = useDisclosure();

  const onUpdateShapeFill = (newShapeFill: number) => updateWidget({ shapeFill: newShapeFill });
  const onUpdateColorOne = (newShapeColorOne: string) => updateWidget({ shapeColorOne: newShapeColorOne });
  const onUpdateColorTwo = (newShapeColorTwo: string) => updateWidget({ shapeColorTwo: newShapeColorTwo });

  const colorFillTriggerProps: DefaultTriggerProps = useMemo(
    () => ({
      onClick: toggleFillColor,
    }),
    [toggleFillColor],
  );

  const colorFillTwoTriggerProps: DefaultTriggerProps = useMemo(
    () => ({
      onClick: toggleFillColorTwo,
    }),
    [toggleFillColorTwo],
  );

  const hasSecondaryColorFill = shapeFill > 0;

  return (
    <>
      <Box
        onClick={() => toggleColorFill()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <IconButton mb='10px' size='sm' aria-label='Font Icon' icon={<FillIcon />} />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Shape Fill %
          </Text>
        </Box>
      </Box>

      <Box
        onClick={() => toggleFillColor()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <DefaultTrigger iconStyle='fill' color={shapeColorOne} {...colorFillTriggerProps} />

        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Color One
          </Text>
        </Box>
      </Box>

      {hasSecondaryColorFill && (
        <Box
          onClick={() => toggleFillColorTwo()}
          mr='5px'
          p='10px 10px'
          as='div'
          style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
        >
          <DefaultTrigger iconStyle='fill' color={shapeColorTwo} {...colorFillTwoTriggerProps} />

          <Box w='75px' as='div'>
            <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
              Fill Color Two
            </Text>
          </Box>
        </Box>
      )}

      <Drawer placement='bottom' onClose={toggleColorFill} isOpen={isColorFillOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Adjust Percentage</p>
              <CloseIcon onClick={toggleColorFill} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <FillPercentOption value={shapeFill} onChange={onUpdateShapeFill} />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>

      <Drawer placement='bottom' onClose={toggleFillColor} isOpen={isFillColorOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Shape Color</p>
              <CloseIcon onClick={toggleFillColor} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <ColorPickerInContext
                iconStyle='text'
                color={shapeColorOne}
                label={'Shape Color'}
                onChange={onUpdateColorOne}
                showNoColorOption={false}
              />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>

      <Drawer placement='bottom' onClose={toggleFillColorTwo} isOpen={isFillColorTwoOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Shape Color Two</p>
              <CloseIcon onClick={toggleFillColorTwo} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <ColorPickerInContext
                iconStyle='text'
                color={shapeColorTwo}
                label={'Shape Color Two'}
                onChange={onUpdateColorTwo}
                showNoColorOption={false}
              />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
};

const FillPercentOption = ({ value, onChange }: { value: number; onChange: (f: number) => void }) => {
  return (
    <SliderNumberInput
      title={'Shape Fill'}
      value={value}
      onChange={onChange}
      suffix={'%'}
      sliderWidth={120}
      containerProps={{ p: 2, pt: 3 }}
    />
  );
};

const DefaultTrigger = ({ iconStyle, color, ...defaultTriggerProps }: Props) => {
  const isBorder = iconStyle === 'border';
  return (
    <Flex
      w={TRIGGER_BUTTON_SIZE}
      h={TRIGGER_BUTTON_SIZE}
      mb='10px'
      borderRadius='5px'
      as='button'
      transition='opacity 0.2s'
      _focus={{
        boxShadow: 'var(--vg-shadows-outline)',
        outline: 'none',
      }}
      _hover={{ bg: 'hover.gray' }}
      justifyContent='center'
      alignItems='center'
      background='#dbdbdb'
      {...defaultTriggerProps}
    >
      <Flex
        w={COLOR_ICON_SIZE}
        h={COLOR_ICON_SIZE}
        borderRadius='sm'
        background={color}
        border='1px solid var(--vg-colors-black-alpha-500)'
        justifyContent='center'
        alignItems='center'
      >
        {isBorder && <Box w={2} h={2} bg='white' boxShadow='0px 0px 0px 1px var(--vg-colors-black-alpha-500) inset' />}
      </Flex>
    </Flex>
  );
};
