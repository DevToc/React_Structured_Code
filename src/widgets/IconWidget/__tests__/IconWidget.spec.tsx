import { renderWidget } from 'utils/test-utils.test';
import { IconWidget } from '../IconWidget';
import { IconWidgetType } from '../IconWidget.types';
import { generateDefaultData } from '../IconWidget.helpers';

describe('widgets/IconWidget', () => {
  it('should render correctly', () => {
    const widgetId = '002.widget-1';
    const widget = { widgetId, ...generateDefaultData('icon-1') };
    const { asFragment } = renderWidget(<IconWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the icon type correctly', () => {
    const widgetId = '002.widget-1';
    const widget = {
      widgetId,
      ...generateDefaultData('icons8-6110', '', undefined, undefined, {
        widthPx: 550,
        heightPx: 42,
        shapeFill: 67,
        shapeColorOne: '#0052A3',
        shapeColorTwo: '#8ABFF4',
        type: IconWidgetType.Grid,
      }),
    };
    const { asFragment } = renderWidget(<IconWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should mark icon as decorative by default', () => {
    const widgetId = '002.widget-1';
    const props = { widgetId, ...generateDefaultData('icon-1') };

    expect(props.widgetData.isDecorative).toBeTruthy();
  });
});
