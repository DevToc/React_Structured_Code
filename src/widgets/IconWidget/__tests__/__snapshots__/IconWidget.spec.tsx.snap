// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/IconWidget should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="002.widget-1-description"
    aria-label="Icon widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-002.widget-1"
    id="002.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 300px; height: 300px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="002.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/IconWidget should render the icon type correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="002.widget-1-description"
    aria-label="Icon widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-002.widget-1"
    id="002.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 550px; height: 42px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="002.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
